### PR TITLE
[feature] new plugin for SPARC spectrum acquisition with arbitrary scan order

### PR DESCRIPTION
**Pull request by pieleric**
Allow to acquire a spectrum acquisition where the beam doesn't scan in
the standard order (ie, fast X, then slow Y). The plugin scans by
default along a checkboard pattern, but the idea is that the user can
change the patter to his/her liking by adding a different function.

### CODE REVIEW

File: plugins/spectrum_arbscor.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        :param px_idx: y, x position
+        :param rep: x, y number of points in teh scan
+        """
+        # px_idx
+        px_idx_flat = px_idx[0] * rep[0] + px_idx[1]
+        act_px_idx = self._px_order[px_idx_flat][::-1]
+        logging.debug("Converted back idx %s to %s", px_idx, act_px_idx)
+        return super()._assembleLiveData(n, raw_data, act_px_idx, rep, pol_idx)
+
+
```
Comment (sneepstefan) [inline plugins/spectrum_arbscor.py:156]: could you please remove 2 blank lines here (PEP 8)?

File: plugins/spectrum_arbscor.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+
+
+class SpectrumArbScOrPlugin(Plugin):
```
Comment (sneepstefan) [inline plugins/spectrum_arbscor.py:159]: class name looks a bit silly, perhaps just change it to SpectrumArbitraryScanOrderPlugin?

File: plugins/spectrum_arbscor.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    name = "Spectrum acquisition in arbitrary scan order"
+    __version__ = "1.0"
+    __author__ = u"Éric Piel"
+    __license__ = "GPLv2"
+
+    def __init__(self, microscope, main_app):
+        super().__init__(microscope, main_app)
+        # Can only be used with a SPARC with spetrometer and CCD camera
```
Comment (sneepstefan) [inline plugins/spectrum_arbscor.py:167]:
```suggestion
        # Can only be used with a SPARC with spectrometer and CCD camera
```

File: plugins/spectrum_arbscor.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+class SEMSpectrumArbitraryOrderMDStream(SEMSpectrumMDStream):
+    """
+    Same as Spectrum stream, but different scan order.
+    """
+
+    def _checkboard_scan_order(self, rep: Tuple[int, int]) -> List[Tuple[int, int]]:
+        """
+        First scan every odd pixel, and then every even pixel. Like first scanning the white
+        squares of a checkboard, and then the black squares.
```
Comment (sneepstefan) [inline plugins/spectrum_arbscor.py:54]: either chessboard or checkerboard

File: plugins/spectrum_arbscor.py [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        raw_order = numpy.ndindex(*rep[::-1])
+        for raw_idx, px_idx in zip(raw_order, order):
+            beam_pos = (pos00[0] + px_idx[0] * pxs[0],
+                        pos00[1] + px_idx[1] * pxs[1],)
+            pos[raw_idx] = beam_pos
+
+        logging.debug("Will acquire at beam pos: %s", pos)
+        return pos
+
+    def _assembleLiveData(self, n, raw_data, px_idx, rep, pol_idx):
```
Comment (sneepstefan) [inline plugins/spectrum_arbscor.py:145]: could you explicitly use typing here in the parameter definition as well?

File: plugins/spectrum_arbscor.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        """
+        :param px_idx: y, x position
+        :param rep: x, y number of points in teh scan
```
Comment (sneepstefan) [inline plugins/spectrum_arbscor.py:148]:
```suggestion
        :param rep: x, y number of points in the scan
```

File: plugins/spectrum_arbscor.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                actname = "Spectrum Arbitrary Scan"
+            else:
+                actname = "Spectrum Arbitrary Scan with %s" % (sptm.name,)
+            act = functools.partial(self.add_stream, name=actname, detector=sptm)
+            stctrl.add_action(actname, act)
+
+    def add_stream(self, name: str, detector: "Detector"):
+        """
+        Create a new spectrum stream and MDStream.
+        :param name:
```
Comment (sneepstefan) [inline plugins/spectrum_arbscor.py:187]: fill in the param name?

**Review by sneepstefan — CHANGES_REQUESTED** [review]

**Review by patrickcleeve2 — APPROVED** [review]

File: plugins/spectrum_arbscor.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+
+    def _checkboard_scan_order(self, rep: Tuple[int, int]) -> List[Tuple[int, int]]:
+        """
+        First scan every odd pixel, and then every even pixel. Like first scanning the white
+        squares of a checkboard, and then the black squares.
+        Example:
+        192.3.4.
+        .5.6.7.8
+        :param rep:
+        :return:
```
[the remark this answers is not in the document]
Comment (pieleric) [reply plugins/spectrum_arbscor.py:59]: Mentioned it's the same as _get_scan_order()

File: plugins/spectrum_arbscor.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+from odemis import model
+from odemis.acq.stream import SEMCCDMDStream, PolarizedCCDSettingsStream, TemporalSpectrumStream, SEMSpectrumMDStream, \
+    SpectrumSettingsStream
+from odemis.gui.conf.data import get_local_vas
+from odemis.gui.plugin import Plugin
+from odemis.model import MD_DESCRIPTION, MD_DIMS, MD_PIXEL_SIZE, MD_POS, MD_ROTATION, MD_ROTATION_COR, MD_WL_LIST
+
+
+# Use a normal settings stream for the spectrum, but a dedicated MDStream that changes the e-beam
+# scan order.
```
[the remark this answers is not in the document]
Comment (pieleric) [reply plugins/spectrum_arbscor.py:44]: This is to explain the way the entire plugin works.... Maybe not very clear. I've improved it.

File: plugins/spectrum_arbscor.py [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+        return order
+
+    def _get_scan_order(self, rep: Tuple[int, int]) -> List[Tuple[int, int]]:
+        """
+        Computes the arbitrary scan order
+        :param rep: x,y number of pixels to scan
+        :return: list of positions (x, y) as indices to scan, in the scan order
+        """
+        #return self._checkboard_scan_order(rep)
```
[the remark this answers is not in the document]
Comment (pieleric) [reply plugins/spectrum_arbscor.py:104]: I'd rather not. The idea is to show that you can use any of the function. I've added a comment explaining that it's to be changed by the user.

File: plugins/spectrum_arbscor.py [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+
+        logging.debug("Will acquire at beam pos: %s", pos)
+        return pos
+
+    def _assembleLiveData(self, n, raw_data, px_idx, rep, pol_idx):
+        """
+        :param px_idx: y, x position
+        :param rep: x, y number of points in teh scan
+        """
+        # px_idx
```
[the remark this answers is not in the document]
Comment (pieleric) [reply plugins/spectrum_arbscor.py:150]: I've extended the explanation in the doctring.

File: plugins/spectrum_arbscor.py (hunk 2 above)
Comment (pieleric) [reply plugins/spectrum_arbscor.py:159]: I was very proud of an abbreviated name that can still be pronounce, and you call it silly! :wink: 
Ah... well I'll put the full name. It's actually fine too.

Comment (sneepstefan) [reply plugins/spectrum_arbscor.py]: Haha yeah well with camelcase notation it looked a bit silly, sorry for that.

**Review by sneepstefan — APPROVED** [review]

**Pull request merged**